File: javascript/18-big-bang-algorithm/test/fraud_detector.test.js
import assert from 'node:assert';
import test from 'node:test';
import { FraudDetector } from '../src/fraud_detector.js';

test('low risk small transaction', () => {
  const detector = new FraudDetector();
  const result = detector.detect({ amount: 10, timestamp: Date.now(), history: [], merchant: 'grocery', country: 'US', cardCountry: 'US' });
  assert.strictEqual(result.rating, 'low');
  assert.strictEqual(result.level, 1);
});

test('medium risk for large amount', () => {
  const detector = new FraudDetector();
  const result = detector.detect({ amount: 1100, timestamp: Date.now(), history: [], merchant: 'grocery', country: 'US', cardCountry: 'US' });
  assert.strictEqual(result.rating, 'medium');
  assert.strictEqual(result.level, 2);
});

test('elevated risk for gambling', () => {
  const detector = new FraudDetector();
  const result = detector.detect({ amount: 100, timestamp: Date.now(), history: [], merchant: 'gambling', country: 'US', cardCountry: 'US' });
  assert.strictEqual(result.rating, 'medium');
  assert.strictEqual(result.level, 2);
});

test('high risk for cross-border', () => {
  const detector = new FraudDetector();
  const result = detector.detect({ amount: 1000, timestamp: Date.now(), history: [], merchant: 'grocery', country: 'FR', cardCountry: 'US' });
  assert.strictEqual(result.rating, 'low');
  assert.strictEqual(result.level, 1);
});

test('critical risk for late night crypto', () => {
  const t = new Date('2024-01-01T02:00:00Z').getTime();
  const detector = new FraudDetector();
  const result = detector.detect({ amount: 600, timestamp: t, history: [], merchant: 'crypto', country: 'CN', cardCountry: 'US' });
  assert.strictEqual(result.rating, 'elevated');
  assert.strictEqual(result.level, 3);
});

test('velocity increases risk', () => {
  const now = Date.now();
  const detector = new FraudDetector();
  const result = detector.detect({ amount: 50, timestamp: now, history: [
    { amount: 10, timestamp: now - 1000 },
    { amount: 10, timestamp: now - 2000 },
    { amount: 10, timestamp: now - 3000 },
    { amount: 10, timestamp: now - 4000 }
  ], merchant: 'grocery', country: 'US', cardCountry: 'US' });
  assert.strictEqual(result.level, 2);
});

test('volume spikes increase risk', () => {
  const now = Date.now();
  const detector = new FraudDetector();
  const result = detector.detect({ amount: 50, timestamp: now, history: [
    { amount: 200, timestamp: now - 10000 },
    { amount: 200, timestamp: now - 20000 },
    { amount: 200, timestamp: now - 30000 }
  ], merchant: 'grocery', country: 'US', cardCountry: 'US' });
  assert.strictEqual(result.level, 1);
});

test('includes score', () => {
  const detector = new FraudDetector();
  const result = detector.detect({ amount: 10, timestamp: Date.now(), history: [], merchant: 'grocery', country: 'US', cardCountry: 'US' });
  assert.strictEqual(typeof result.score, 'number');
});

File: csharp/18-big-bang-algorithm/test/fraud_detector_test.cs
using Xunit;

public class FraudDetectorTest
{
    private long Now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    private Tx CreateTx(decimal amount, long timestamp, List<Tx>? history = null, string merchant = "grocery", string country = "US", string cardCountry = "US")
    {
        return new Tx(amount, timestamp, history ?? new List<Tx>(), merchant, country, cardCountry);
    }

    [Fact]
    public void LowRiskSmallTransaction()
    {
        var detector = new FraudDetector();
        var result = detector.detect(CreateTx(10, Now));
        Assert.Equal("low", result.Rating);
        Assert.Equal(1, result.Level);
    }

    [Fact]
    public void MediumRiskForLargeAmount()
    {
        var detector = new FraudDetector();
        var result = detector.detect(CreateTx(1100, Now));
        Assert.Equal("medium", result.Rating);
        Assert.Equal(2, result.Level);
    }

    [Fact]
    public void GamblingMerchantIsMediumRisk()
    {
        var detector = new FraudDetector();
        var result = detector.detect(CreateTx(100, Now, merchant: "gambling"));
        Assert.Equal("medium", result.Rating);
        Assert.Equal(2, result.Level);
    }

    [Fact]
    public void CrossBorderAloneIsLowRisk()
    {
        var detector = new FraudDetector();
        var result = detector.detect(CreateTx(1000, Now, country: "FR", cardCountry: "US"));
        Assert.Equal("low", result.Rating);
        Assert.Equal(1, result.Level);
    }

    [Fact]
    public void CriticalRiskForLateNightCrypto()
    {
        var detector = new FraudDetector();
        var t = new DateTimeOffset(2024, 1, 1, 2, 0, 0, TimeSpan.Zero).ToUnixTimeMilliseconds();
        var result = detector.detect(CreateTx(600, t, merchant: "crypto", country: "CN", cardCountry: "US"));
        Assert.Equal("elevated", result.Rating);
        Assert.Equal(3, result.Level);
    }

    [Fact]
    public void VelocityIncreasesRisk()
    {
        var detector = new FraudDetector();
        var history = new List<Tx>
        {
            CreateTx(10, Now - 1000, merchant: "", country: "", cardCountry: ""),
            CreateTx(10, Now - 2000, merchant: "", country: "", cardCountry: ""),
            CreateTx(10, Now - 3000, merchant: "", country: "", cardCountry: ""),
            CreateTx(10, Now - 4000, merchant: "", country: "", cardCountry: "")
        };
        var result = detector.detect(CreateTx(50, Now, history));
        Assert.Equal(2, result.Level);
    }

    [Fact]
    public void VolumeSpikesAloneStayLowRisk()
    {
        var detector = new FraudDetector();
        var history = new List<Tx>
        {
            CreateTx(200, Now - 10000, merchant: "", country: "", cardCountry: ""),
            CreateTx(200, Now - 20000, merchant: "", country: "", cardCountry: ""),
            CreateTx(200, Now - 30000, merchant: "", country: "", cardCountry: "")
        };
        var result = detector.detect(CreateTx(50, Now, history));
        Assert.Equal(1, result.Level);
    }

    [Fact]
    public void IncludesScore()
    {
        var detector = new FraudDetector();
        var result = detector.detect(CreateTx(10, Now));
        Assert.IsType<int>(result.Score);
    }
}
